### PR TITLE
314 create llm corpus

### LINKED ISSUE #314: Create LLM corpus
## Issue description

Some LLMs are currently part of the `popular_on_huggingface` corpus of MLAgility. Those models are significantly large, causing our benchmarking infrastructure to take significantly longer when processing them. 

## Task

Create corpus composed only of LLM models, since those models must be treated differently than most models.

## Suggested implementation

* Move all LLMs from `popular_on_huggingface` to a corpus called `LLMs`
* Add new popular LLMs that are not already part of the `popular_on_huggingface` corpus
* Create models from `configs` rather than loading pre-trained models to avoid long download times.

**Example:**

```
from transformers import OpenAIGPTModel, OpenAIGPTConfig
model = OpenAIGPTModel(config=OpenAIGPTConfig()) 
```

Instead of 

```
from transformers import OpenAIGPTModel
model = OpenAIGPTModel.from_pretrained("openai-gpt")
```

### Decoder-Only

* 2023
  * LLaMA (Meta)
* 2022
  * OPT-IML (Meta)
  * BLOOM (BigScience)
  * BLOOMZ (BigScience)
  * Galactica (Meta)
  * OPT (Meta)
  * YaLM (Yandex)
  * GPT-NeoX (EleutherAI)
* 2021
  * GPT-J (EleutherAI)
  * GPT-Neo (EleutherAI)
* 2019
  * XLNet (Google)
  * GPT-2 (OpenAI)
* 2018
  * GPT-1 (OpenAI)
  
### Encoder-Decoder

* 2023
  * Chat GLM (THUDM)
  * Flan UL2 (Google)
* 2022
  * Flan T5 (Google)
  * UL2 (Google)
  * Tk (Huggingface)
* 2021
  * GLM (THUDM)
  * Switch (Google)
* 2020
  * mT5 (Google)
  * T0 (BigScience)
* 2019
  * T5 (Google)
  * BART (Meta)

### Encoder-Only

* 2020
  * Deberta (Microsoft)
  * ELECTRA (Stanford)
* 2019
  * RoBERTa (Meta)
  * ALBERT (Google)
  * DistillBERT (Huggingface) - Not an LLM?
  * ERNIE (Baidu)
* 2018
  * BERT (Google)
  
## Suggested list of mandatory labels for LLMs

* `arch::encoder_only,decoder_only,encoder_decoder`
* `author::<>`
* `year::<>`

### PR DESCRIPTION
**Pull request by jeremyfowers**
Closes #314 

## Overview

Adds two new corpora: `llm` and `llm_layer`. Both corpora feature `gpt-neo`, `gpt-j`, and `gpt-neox`. `llm_layer` also includes 8 LLaMA layers, corresponding to the 4 supported LLaMA model sizes with KV-caching disabled and enabled. 

I am sure we will add more models to these corpora in the future, however this seems like a good start to unblock other work.

Makes the following changes in mlagility source code to enable LLaMA support:
- Changes `benchit cache list` to print in a single column, since the model names are getting rather long and the names overlap with each other if we print multiple columns.
- Changes `mlagility.analysis.get_model_hash()` to consider model parameters, to work around bug #316 
  - Updates a hardcoded hash in CLI test 004
  - Updates CLI test 015 to no longer rely on hardcoded hashes
- Removes `detoxify` from `models/requirements.txt` and 3 models from `models/popular_on_huggingface` that relied on `detoxify`. This was necessary because `detoxify` is an abandoned library that was preventing us from updating to the latest versions of `transformers`.
- Changes `onnxflow.common.build.get_shapes_and_dtypes()` to support `List[torch.tensor]`, which is used in LLaMA's KV-caching

## Demo

The command:
`benchit models/llm_layer/llama*.py --build-only`

Produces results:
```
Models discovered during profiling:

llama_13b_cache_layer.py:
        model (executed 1x)
                Model Type:     Pytorch (torch.nn.Module)
                Class:          LlamaDecoderLayer (<class 'transformers.models.llama.modeling_llama.LlamaDecoderLayer'>)
                Location:       /net/home/jfowers/Groq/mlagility/models/llm_layer/llama_layer_prototype.py, line 62
                Parameters:     419,440,640 (800.0 MB)
                Hash:           395eda51
                Status:         Model successfully built!

llama_13b_layer.py:
        model (executed 1x)
                Model Type:     Pytorch (torch.nn.Module)
                Class:          LlamaDecoderLayer (<class 'transformers.models.llama.modeling_llama.LlamaDecoderLayer'>)
                Location:       /net/home/jfowers/Groq/mlagility/models/llm_layer/llama_layer_prototype.py, line 95
                Parameters:     419,440,640 (800.0 MB)
                Hash:           7680d2d8
                Status:         Model successfully built!

llama_33b_cache_layer.py:
        model (executed 1x)
                Model Type:     Pytorch (torch.nn.Module)
                Class:          LlamaDecoderLayer (<class 'transformers.models.llama.modeling_llama.LlamaDecoderLayer'>)
                Location:       /net/home/jfowers/Groq/mlagility/models/llm_layer/llama_layer_prototype.py, line 62
                Parameters:     708,850,688 (1352.0 MB)
                Hash:           fea39109
                Status:         Model successfully built!

llama_33b_layer.py:
        model (executed 1x)
                Model Type:     Pytorch (torch.nn.Module)
                Class:          LlamaDecoderLayer (<class 'transformers.models.llama.modeling_llama.LlamaDecoderLayer'>)
                Location:       /net/home/jfowers/Groq/mlagility/models/llm_layer/llama_layer_prototype.py, line 95
                Parameters:     708,850,688 (1352.0 MB)
                Hash:           827f0abd
                Status:         Model successfully built!

llama_65b_cache_layer.py:
        model (executed 1x)
                Model Type:     Pytorch (torch.nn.Module)
                Class:          LlamaDecoderLayer (<class 'transformers.models.llama.modeling_llama.LlamaDecoderLayer'>)
                Location:       /net/home/jfowers/Groq/mlagility/models/llm_layer/llama_layer_prototype.py, line 62
                Parameters:     1,073,758,208 (2048.0 MB)
                Hash:           25375a7d
                Status:         Model successfully built!

llama_65b_layer.py:
        model (executed 1x)
                Model Type:     Pytorch (torch.nn.Module)
                Class:          LlamaDecoderLayer (<class 'transformers.models.llama.modeling_llama.LlamaDecoderLayer'>)
                Location:       /net/home/jfowers/Groq/mlagility/models/llm_layer/llama_layer_prototype.py, line 95
                Parameters:     1,073,758,208 (2048.0 MB)
                Hash:           8521b102
                Status:         Model successfully built!

llama_7b_cache_layer.py:
        model (executed 1x)
                Model Type:     Pytorch (torch.nn.Module)
                Class:          LlamaDecoderLayer (<class 'transformers.models.llama.modeling_llama.LlamaDecoderLayer'>)
                Location:       /net/home/jfowers/Groq/mlagility/models/llm_layer/llama_layer_prototype.py, line 62
                Parameters:     268,443,648 (512.0 MB)
                Hash:           e2debc06
                Status:         Model successfully built!

llama_7b_layer.py:
        model (executed 1x)
                Model Type:     Pytorch (torch.nn.Module)
                Class:          LlamaDecoderLayer (<class 'transformers.models.llama.modeling_llama.LlamaDecoderLayer'>)
                Location:       /net/home/jfowers/Groq/mlagility/models/llm_layer/llama_layer_prototype.py, line 95
                Parameters:     268,443,648 (512.0 MB)
                Hash:           a4c6ff51
                Status:         Model successfully built!


Woohoo! The 'benchmark' command is complete.
```

### CODE REVIEW

File: models/llm_layer/llama_layer_prototype.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
+# labels: name::llama_7b_cache_layer author::transformers task::Natural_Language_Processing
```
Comment (ramkrishna2910) [inline models/llm_layer/llama_layer_prototype.py:1]: Should LLaMa fall under `Generative_AI` tag?
Or we should start supporting multiple tags per model.

File: src/mlagility/analysis/util.py [hunk 2]
```diff
@@ -91,11 +91,21 @@ def populate_onnx_model_info(onnx_model) -> Dict:
...
+            {
+                "size on disk (KiB)": round(
+                    model.SerializeToString().__sizeof__() / 1024, 4
+                ),
+            }
+        )
+    except ValueError:
+        # Models >2GB on disk cannot have their model size measured this
+        # way and will throw a ValueError
+        pass
```
Comment (ramkrishna2910) [inline src/mlagility/analysis/util.py:107]: Can we add an issue to handle this?
Currently, the fields will be empty which is not ideal.

**Review by ramkrishna2910 — APPROVED** [review]
Just added a couple of minor comments. Looks good!

File: src/mlagility/analysis/util.py (hunk 2 above)
Comment (jeremyfowers) [reply src/mlagility/analysis/util.py]: filed https://github.com/groq/mlagility/issues/318

File: models/llm_layer/llama_layer_prototype.py (hunk 1 above)
Comment (jeremyfowers) [reply models/llm_layer/llama_layer_prototype.py]: https://github.com/groq/mlagility/pull/317/commits/4dc27894b0f92739d9dafdfaa8031f39b42cb9ba

File: src/mlagility/analysis/util.py (hunk 2 above)
Comment (jeremyfowers) [reply src/mlagility/analysis/util.py]: https://github.com/groq/mlagility/pull/317/commits/de2f36182ad8bfffa05d67aa0c3ea605706da6ba
